site-src/geps/gep-1058.md

```diff
@@ -0,0 +1,962 @@
+
+In this case, each controller is expected only to update its own entry in `parents`.
+
+```yaml
+---
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: Gateway
+metadata:
+  name: gateway-a
+spec:
+  gatewayClassName: acme-lb
+  listeners:
+  - name: http
+    protocol: HTTP
+    port: 80
+    hostname: *.example.com
+---
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: Gateway
+metadata:
+  name: gateway-b
+spec:
+  gatewayClassName: nadir-lb
+  listeners:
+  - name: http
+    protocol: HTTP
+    port: 80
+    hostname: subdomain.example.com
+---
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: HTTPRoute
+metadata:
+  name: http-parent
+spec:
+  parentRefs:
+  - name: gateway-a
+    sectionName: http
+  - name: gateway-b
+    sectionName: http
+  rules:
+    - allowedRoutes:
+        kinds: 
+          - kind: HTTPRoute
+        namespaces:
+          from: Same
+    - backendRefs:
+      - name: default-svc
+        port: 8080
+status:
+  parents:
+  - controllerName: acme
+    parentRef:
+      name: gateway-a
+    conditions:
+    - type: Accepted
+      status: true
+      reason: Accepted
+      message: Successfully accepted
+  - controllerName: nadir
+    parentRef:
+      name: gateway-b
+    conditions:
+    - type: Accepted
+      status: true
+      reason: Accepted
+      message: Successfully accepted
+---
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: HTTPRoute
+metadata:
+  name: http-child
+spec:
+  parentRefs:
+  - name: gateway-a
+    sectionName: http
+  hostnames:
+    - testdomain.example.com
+  rules:
+    - backendRefs:
+      - name: foo-svc
+        port: 8080
+status:
+  parents:
+  - controllerName: acme
+    parentRef:
+      name: http-parent
+    conditions:
+    - type: Accepted
+      status: true
+      reason: Accepted
+      message: Accepted successfully
+    grandparents:
+      - parentRef:
+          name: gateway-a
+        conditions:
+        - type: Accepted
+          status: true
+          reason: Accepted
+          message: Successfully accepted
+  - controllerName: nadir
+    parentRef:
+      name: http-parent
+    conditions:
+    - type: Accepted
+      status: true
+      reason: GrandparentConfigError
+      message: Some configuration was not accepted in the attachment process. See grandparents for more details.
+    grandparents:
+      - parentRef:
+          name: gateway-b
+        conditions:
+        - type: Accepted
+          status: false
+          reason: HostnameMismatch
+          message: The route could not be accepted because of a hostname mismatch - `testdomain.example.com` does not match `subdomain.example.com`.
+```
+
+
+
+## Alternatives
+
+### Extending backendRef to allow targeting a Route
+
+This idea was to allow `backendRef` to, optionally, specify a route instead of a
+service. Some of the reasons for not taking this approach include:
+
+- It would require changes to `backendRefs`. No changes to `allowedRoutes` are
```

Comment:
I disagree. backendRef can refer to any type, including route. Quite the opposite - this proposal requires changes to all route types to add `allowedRoutes`

Reply:
Another way to look at this would be that route delegation may only make sense for a subset of routes, maybe just HTTPRoute to start. I personally don't like using backendRef for this concept because another route is not really a backend, it's just additional routing configuration. 

Maybe more importantly, the separation of concerns here seems like a critical part of the design. I want to be able to say that I trust Route admins in the foo namespace to handle routing for anything under `/foo`. I shouldn't have to care about what the specific Route(s) are called in that namespace.

Reply:
Agree with the last point - and we should match whatever GAMMA does. 

For example, if we go with 'attach routes to Services' - then the backend remains a Service ( no change ), and we can delegate to the HttpRoute attached to the service in the child namespace. 

If we add a new CRD to represent a virtual service - this can be the backendRef. If mesh would treat this virtual service with attached routes as a backend - so should the gateway spec. 

Service and whatever resource GAMMA defines will be named entities - you shouldn't care how the HttpRoute is named, but the name of the Service or virtual Service is reflected in DNS and exposed to users.

A nice side-effect is that the 'child' routes could be directly used from the mesh - don't know if we want to allow this, see my comment on interactions between gateways and mesh.